### PR TITLE
[mojo-stdlib] Replace `register_passable __init__ -> Self` usages in `stdlib/src/builtin`

### PR DESCRIPTION
**Pull request by MertCelik**
Replaces `register_passable __init__ -> Self` usages in `stdlib/src/builtin`. First of 6 PRs for https://github.com/modularml/mojo/issues/2037

- [x] Replace `/builtin` #2075
- [ ] Replace `/collections`
- [ ] Replace `/memory`
- [ ] Replace `/os`
- [ ] Replace `/utils`
- [ ] Replace misc

### CODE REVIEW

File: stdlib/src/builtin/int_literal.mojo [hunk 1]
```diff
@@ -35,13 +35,13 @@ struct IntLiteral(Intable, Stringable, Boolable, EqualityComparable):
...
 
     @always_inline("nodebug")
-    fn __init__() -> Self:
+    fn __init__(inout self):
         """Default constructor.
 
         Returns:
             The constructed Self object.
         """
-        return Self._zero
```
Comment (MertCelik) [inline stdlib/src/builtin/int_literal.mojo:44]: Not entirely comfortable with this usage, is it correct?

File: stdlib/src/builtin/int_literal.mojo [hunk 2]
```diff
@@ -35,13 +35,13 @@ struct IntLiteral(Intable, Stringable, Boolable, EqualityComparable):
...
+        self._zero
```
Comment (MertCelik) [inline stdlib/src/builtin/int_literal.mojo:44]: Not entirely comfortable with this usage, is it correct?

File: stdlib/src/collections/optional.mojo [hunk 3]
```diff
@@ -243,13 +243,13 @@ struct OptionalReg[T: AnyRegType](Boolable):
...
 
-    fn __init__() -> Self:
+    fn __init__(inout self):
         """Create an optional without a value.
 
         Returns:
             The optional.
         """
-        return Self(None)
+        self(None)
```
Comment (MertCelik) [inline stdlib/src/collections/optional.mojo:252]: ditto

File: stdlib/src/builtin/int_literal.mojo (hunk 2 above)
Comment (lattner) [reply stdlib/src/builtin/int_literal.mojo:44]: No, this isn't correct, you need `self = Self._zero`.

File: stdlib/src/builtin/int_literal.mojo [hunk 4]
```diff
@@ -35,13 +35,13 @@ struct IntLiteral(Intable, Stringable, Boolable, EqualityComparable):
     )
 
     @always_inline("nodebug")
-    fn __init__() -> Self:
+    fn __init__(inout self):
         """Default constructor.
 
         Returns:
             The constructed Self object.
```
Comment (lattner) [inline stdlib/src/builtin/int_literal.mojo:42]: This will need to remove the "returns" doc comment, since it is nuking the return value.

**Review by lattner — COMMENTED** [review]
Thank you for working on this. It would be good to carve this up into a few smaller patches, could you do one directory at a time?   Please also split the changes to 'unsafe.mojo' out, those have some internal issues and I'll take care of it.

**Comment by MertCelik** [comment]
> Thank you for working on this. It would be good to carve this up into a few smaller patches, could you do one directory at a time? Please also split the changes to 'unsafe.mojo' out, those have some internal issues and I'll take care of it.

I reduced the scope of this PR to `stdlib/src/builtin` only, updated the description and addressed the change requests. 🚀

File: stdlib/src/builtin/object.mojo [hunk 5]
```diff
@@ -340,8 +340,8 @@ struct _ObjectImpl(CollectionElement, Stringable):
         return Self {value: value}
 
     @always_inline
-    fn __init__() -> Self:
-        return __mlir_op.`kgen.variant.create`[
+    fn __init__(inout self):
+        self = __mlir_op.`kgen.variant.create`[
```
Comment (MertCelik) [inline stdlib/src/builtin/object.mojo:344]: this turned out trickier than expected, need more reading

**Comment by JoeLoser** [comment]
Looking good, but you'll need to fix up the formatting.  This can be done with `mojo format -l 80 stdlib/src/builtin/simd.mojo` in your particular case.  Thanks!

**Comment by MertCelik** [comment]
> Looking good, but you'll need to fix up the formatting. This can be done with `mojo format -l 80 stdlib/src/builtin/simd.mojo` in your particular case. Thanks!

I think due to rebase, it created some issue with git diff action
image
https://github.com/modularml/mojo/actions/runs/8539518017/job/23394278326
I formatted all the files for 80 but somehow still failing the workflow

**Comment by JoeLoser** [comment]
I marked this PR with the "hold off" label - see my comment in [the issue](https://github.com/modularml/mojo/issues/2037#issuecomment-2036161910).  Once we get a chance to take a look at the performance difference issues internally, we will circle back and can continue working on this.  Sorry for the inconvenience here as the performance difference was a bit surprising.

**Comment by gabrieldemarmiesse** [comment]
The codebase changed quite a lot since the PR was opened and it seems that there are a lot of conflicts now. @MertCelik would it be worth it to close this pull request and open another one when [this](https://github.com/modularml/mojo/issues/2037#issuecomment-2036161910) has been resolved?

**Comment by laszlokindrat** [comment]
@JoeLoser I believe the special constructor form for reg-passables has been retired. Any objections to closing this?

**Comment by helehex** [comment]
Seems that all remaining cases of this in the stdlib have been updated, so this can be closed.

**Comment by laszlokindrat** [comment]
@MertCelik Thanks again for the patch! Looks like these issues have been addressed elsewhere, but we're looking forward to more contributions from you!

**Pull request closed without merging**